numba_cuda/numba/cuda/cudadrv/managed_module.py

```diff
@@ -0,0 +1,93 @@
+import weakref
+
+from . import devices
+from .driver import CtypesModule, USE_NV_BINDING
+
+
+class _CuFuncProxy:
+    """See `ManagedModule` for more details
+    """
+    def __init__(self, module, cufunc):
+        self._module = module
+        self._cufunc = cufunc
+
+    def init_module(self, stream):
+        self._module._init(stream)
+
+    def lazy_finalize_module(self, stream):
+        self._module._lazy_finalize(stream)
+
+    def __getattr__(self, name):
+        return getattr(self._cufunc, name)
+
+
+class ManagedModule:
+    def __init__(self, module, setup_callbacks, teardown_callbacks):
+        """ctypes module with setup and teardown callbacks
+        The use of managedmodule is the same as a ctypes module,
+        with the exception of `get_function`, which returns a wrapped
+        cufunc object. The wrapped object provides `init_module`
+        and `lazy_finalize_module` method. Which are used
+        to initialize and finalize the module when stream is available
+        in the later stage of the compilation pipeline.
+        """
+        # To be updated to object code
+        if not isinstance(module, CtypesModule):
+            raise TypeError("module must be a CtypesModule")
+
+        if not isinstance(setup_callbacks, list):
+            raise TypeError("setup_callbacks must be a list")
+        if not isinstance(teardown_callbacks, list):
+            raise TypeError("teardown_callbacks must be a list")
+
+        for callback in setup_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+        for callback in teardown_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+
+        self._initialized = False
+        self._module = module
+        self._setup_callbacks = setup_callbacks
+        self._teardown_callbacks = teardown_callbacks
+
+    def _init(self, stream):
+        """Eagerly call the setup functions for cumodule on `stream`
+        """
+        for setup in self._setup_callbacks:
+            setup(self._module, stream)
+
+    def _lazy_finalize(self, stream):
+        """Set teardown function for cumodule via weakref finalizer.
+        """
+        def lazy_callback(callbacks, module, stream):
+            for teardown in callbacks:
+                teardown(module, stream)
+
+        ctx = devices.get_context()
+        if USE_NV_BINDING:
+            key = self._module.handle
+        else:
+            key = self._module.handle.value
+        module_obj = ctx.modules.get(key, None)
+
+        if module_obj is not None:
+            weakref.finalize(
```

Comment:
I spent a while trying to refresh my memory of what we do with finalizers in Numba - https://github.com/numba/numba/pull/10001 may be worth a quick look in conjunction with thinking about the behaviour here.

Reply:
It even took me 30min to understand this refresh PR. I think this is super subtle. Especially the point "the first finalizer registered also calls `atexit.register` to place a function that invokes all finalizers." Is this documented anywhere or just a CPython detail?

It's up to the user to decide whether to inspect the `shutting_down` variable. If they decide to check that, it's probably a good idea to put a reminder about this subtlety. For the most part, if the finalizer is to operate on just the modules, we are probably fine since they are still alive when the finalizer is invoked.